File: lua/darts/darts.lua
local Darts = {}

local scores = {
    outside = 0,
    outer = 1,
    middle = 5,
    inner = 10
}
local radius = {
    inner = 1,
    middle = 5,
    outer = 10
}

function Darts.score(x, y)
    local distance_from_center = math.sqrt(x ^ 2 + y ^ 2)

    if distance_from_center <= radius.inner then
        return scores.inner
    elseif distance_from_center <= radius.middle then
        return scores.middle
    elseif distance_from_center <= radius.outer then
        return scores.outer
    else
        return scores.outside
    end
end

return Darts

File: go/darts/darts.go
package darts

import "math"

// Score calculates the score of the throw according to the area of ​
// the target the dart hit. The deeper the area, the higher the score.
func Score(x, y float64) int {
	distance := distanceFromCenter(x, y)

	for _, zone := range targetZones {
		if float64(zone.size) >= distance {
			return zone.score
		}
	}

	return scoreForMissingTarget
}

// distanceFromCenter calculates the distance at which the point given
// by (x, y) is from the center point (0, 0).
func distanceFromCenter(x, y float64) float64 {
	return math.Sqrt(x*x + y*y)
}

const scoreForMissingTarget int = 0

// targetZones gathers all the zones that the target has:
// inner, middle and outer
var targetZones = []targetZone{
	{1, 10},
	{5, 5},
	{10, 1},
}

// targetZone represents a zone of the target, grouping the size and
// score of the zone.
type targetZone struct {
	size, score int
}
